File: Ui/Qt5.5.1/CustomWindow/Widget.cpp
void drawShadow(QPainter &_painter, qint16 _margin, qreal _radius,
QColor _start, QColor _end, qreal _startPosition, qreal _endPosition0, qreal _endPosition1, qreal _width, qreal _height) {
	_painter.setPen(Qt::NoPen);

	QLinearGradient gradient;
	gradient.setColorAt(_startPosition, _start);
	gradient.setColorAt(_endPosition0, _end);
	// Right
	QPointF right0(_width - _margin, _height / 2);
	QPointF right1(_width, _height / 2);
	gradient.setStart(right0);
	gradient.setFinalStop(right1);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(QPointF(_width - _margin*_radius, _margin), QPointF(_width, _height - _margin)), 0.0, 0.0);
	// Left
	QPointF left0(_margin, _height / 2);
	QPointF left1(0, _height / 2);
	gradient.setStart(left0);
	gradient.setFinalStop(left1);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(QPointF(_margin *_radius, _margin), QPointF(0, _height - _margin)), 0.0, 0.0);
	// Top
	QPointF top0(_width / 2, _margin);
	QPointF top1(_width / 2, 0);
	gradient.setStart(top0);
	gradient.setFinalStop(top1);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(QPointF(_width - _margin, 0), QPointF(_margin, _margin)), 0.0, 0.0);
	// Bottom
	QPointF bottom0(_width / 2, _height - _margin);
	QPointF bottom1(_width / 2, _height);
	gradient.setStart(bottom0);
	gradient.setFinalStop(bottom1);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(QPointF(_margin, _height - _margin), QPointF(_width - _margin, _height)), 0.0, 0.0);
	// BottomRight
	QPointF bottomright0(_width - _margin, _height - _margin);
	QPointF bottomright1(_width, _height);
	gradient.setStart(bottomright0);
	gradient.setFinalStop(bottomright1);
	gradient.setColorAt(_endPosition1, _end);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(bottomright0, bottomright1), 0.0, 0.0);
	// BottomLeft
	QPointF bottomleft0(_margin, _height - _margin);
	QPointF bottomleft1(0, _height);
	gradient.setStart(bottomleft0);
	gradient.setFinalStop(bottomleft1);
	gradient.setColorAt(_endPosition1, _end);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(bottomleft0, bottomleft1), 0.0, 0.0);
	// TopLeft
	QPointF topleft0(_margin, _margin);
	QPointF topleft1(0, 0);
	gradient.setStart(topleft0);
	gradient.setFinalStop(topleft1);
	gradient.setColorAt(_endPosition1, _end);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(topleft0, topleft1), 0.0, 0.0);
	// TopRight
	QPointF topright0(_width - _margin, _margin);
	QPointF topright1(_width, 0);
	gradient.setStart(topright0);
	gradient.setFinalStop(topright1);
	gradient.setColorAt(_endPosition1, _end);
	_painter.setBrush(QBrush(gradient));
	_painter.drawRoundRect(QRectF(topright0, topright1), 0.0, 0.0);
	// Widget
	_painter.setBrush(QBrush("#FFFFFF"));
	_painter.setRenderHint(QPainter::Antialiasing);
	_painter.drawRoundRect(QRectF(QPointF(_margin, _margin), QPointF(_width - _margin, _height - _margin)), _radius, _radius);
}


Widget::Widget() :
_cursorchanged(false),
_leftButtonPressed(false),
_radius(5.0),
_mousePress(None),
_mouseMove(None),
_borderWidth(10)
{
	setMouseTracking(true);
	setWindowFlags(Qt::FramelessWindowHint | Qt::CustomizeWindowHint | Qt::Window);
	setAttribute(Qt::WA_Hover);
	setAttribute(Qt::WA_TranslucentBackground);
	installEventFilter(this);
	_rubberband = new QRubberBand(QRubberBand::Rectangle);
}

Widget::~Widget() {}

void Widget::paintEvent(QPaintEvent *) {
	QPainter painter(this);
	drawShadow(painter, _borderWidth, _radius, QColor(120, 120, 120, 32), QColor(255, 255, 255, 0), 0.0, 1.0, 0.6, width(), height());
}

bool Widget::eventFilter(QObject *o, QEvent*e) {
	if (e->type() == QEvent::MouseMove ||
		e->type() == QEvent::HoverMove ||
		e->type() == QEvent::Leave ||
		e->type() == QEvent::MouseButtonPress ||
		e->type() == QEvent::MouseButtonRelease) {

		switch (e->type())
		{
		case QEvent::MouseMove:
			mouseMove(static_cast<QMouseEvent*>(e));
			return true;
			break;
		case QEvent::HoverMove:
			mouseHover(static_cast<QHoverEvent*>(e));
			return true;
			break;
		case QEvent::Leave:
			mouseLeave(e);
			return true;
			break;
		case QEvent::MouseButtonPress:
			mousePress(static_cast<QMouseEvent*>(e));
			return true;
			break;
		case QEvent::MouseButtonRelease:
			mouseRealese(static_cast<QMouseEvent*>(e));
			return true;
			break;
		}
	}
	else {
		return false;
	}
}

void Widget::mouseHover(QHoverEvent *e) {
	updateCursorShape(mapToGlobal(e->pos()));

}

void Widget::mouseLeave(QEvent *e) {
	Q_UNUSED(e);
	if (!_leftButtonPressed) {
		unsetCursor();
	}
}

void Widget::mousePress(QMouseEvent *e) {
	if (e->button() == Qt::LeftButton) {
		_leftButtonPressed = true;
		calculateCursorPosition(e->globalPos(), frameGeometry(), _mousePress);
		if (_mousePress != None) {
			_rubberband->setGeometry(frameGeometry());
		}
	}
}

void Widget::mouseRealese(QMouseEvent *e) {
	if (e->button() == Qt::LeftButton) {
		_leftButtonPressed = false;
	}
}

void Widget::mouseMove(QMouseEvent *e) {
	if (_leftButtonPressed)
	{
		if (_mousePress != None)
		{
			QRect origRect = _rubberband->frameGeometry();
			int left = origRect.left();
			int top = origRect.top();
			int right = origRect.right();
			int bottom = origRect.bottom();
			if (_mousePress == Top) {
				top = e->globalPos().y();
			}
			else if (_mousePress == Bottom) {
				bottom = e->globalPos().y();
			}
			else if (_mousePress == Left) {
				left = e->globalPos().x();
			}
			else if (_mousePress == Right) {
				right = e->globalPos().x();
			}
			else if (_mousePress == TopLeft) {
				top = e->globalPos().y();
				left = e->globalPos().x();
			}
			else if (_mousePress == TopRight) {
				right = e->globalPos().x();
				top = e->globalPos().y();
			}
			else if (_mousePress == BottomLeft) {
				bottom = e->globalPos().y();
				left = e->globalPos().x();
			}
			else if (_mousePress == BottomRight) {
				bottom = e->globalPos().y();
				right = e->globalPos().x();
			}
			QRect newRect(QPoint(left, top), QPoint(right, bottom));
			if (newRect.width() < minimumWidth() ) {
				left = frameGeometry().x();
			}
			else if (newRect.height() < minimumHeight()  )
			{
				top = frameGeometry().y();
			}
			setGeometry(QRect(QPoint(left, top), QPoint(right, bottom)));
			_rubberband->setGeometry(QRect(QPoint(left, top), QPoint(right, bottom)));
		}
	}
	else {
		updateCursorShape(e->globalPos());
	}
}

void Widget::updateCursorShape(const QPoint &pos) {
	if (isFullScreen() || isMaximized()) {
		if (_cursorchanged) {
			unsetCursor();
		}
		return;
	}
	if (!_leftButtonPressed)
	{
		calculateCursorPosition(pos, frameGeometry(), _mouseMove);
		if (_mouseMove == Top || _mouseMove == Bottom) {
			setCursor(Qt::SizeVerCursor);
			_cursorchanged = true;
		}
		else if (_mouseMove == Left || _mouseMove == Right) {
			setCursor(Qt::SizeHorCursor);
			_cursorchanged = true;
		}
		else if (_mouseMove == TopLeft || _mouseMove == BottomRight) {
			setCursor(Qt::SizeFDiagCursor);
			_cursorchanged = true;
		}
		else if (_mouseMove == TopRight || _mouseMove == BottomLeft) {
			setCursor(Qt::SizeBDiagCursor);
			_cursorchanged = true;
		}
		else if (_cursorchanged) {
			unsetCursor();
			_cursorchanged = false;
		}
	}
}

void Widget::calculateCursorPosition(const QPoint &pos, const QRect &framerect, Edge &_edge) {
	bool onLeft = pos.x() >= framerect.x() - _borderWidth && pos.x() <= framerect.x() + _borderWidth &&
		pos.y() <= framerect.y() + framerect.height() - _borderWidth && pos.y() >= framerect.y() + _borderWidth;

	bool onRight = pos.x() >= framerect.x() + framerect.width() - _borderWidth && pos.x() <= framerect.x() + framerect.width() &&
		pos.y() >= framerect.y() + _borderWidth && pos.y() <= framerect.y() + framerect.height() - _borderWidth;

	bool onBottom = pos.x() >= framerect.x() + _borderWidth && pos.x() <= framerect.x() + framerect.width() - _borderWidth  &&
		pos.y() >= framerect.y() + framerect.height() - _borderWidth && pos.y() <= framerect.y() + framerect.height();

	bool onTop = pos.x() >= framerect.x() + _borderWidth && pos.x() <= framerect.x() + framerect.width() - _borderWidth &&
		pos.y() >= framerect.y() && pos.y() <= framerect.y() + _borderWidth;

	bool  onBottomLeft = pos.x() <= framerect.x() + _borderWidth && pos.x() >= framerect.x() &&
		pos.y() <= framerect.y() + framerect.height() && pos.y() >= framerect.y() + framerect.height() - _borderWidth;

	bool onBottomRight = pos.x() >= framerect.x() + framerect.width() - _borderWidth && pos.x() <= framerect.x() + framerect.width() &&
		pos.y() >= framerect.y() + framerect.height() - _borderWidth && pos.y() <= framerect.y() + framerect.height();

	bool onTopRight = pos.x() >= framerect.x() + framerect.width() - _borderWidth && pos.x() <= framerect.x() + framerect.width() &&
		pos.y() >= framerect.y() && pos.y() <= framerect.y() + _borderWidth;

	bool onTopLeft = pos.x() >= framerect.x() && pos.x() <= framerect.x() + _borderWidth &&
		pos.y() >= framerect.y() && pos.y() <= framerect.y() + _borderWidth;

	if (onLeft) {
		_edge = Left;
	}
	else if (onRight) {
		_edge = Right;
	}
	else if (onBottom) {
		_edge = Bottom;
	}
	else if (onTop) {
		_edge = Top;
	}
	else if (onBottomLeft) {
		_edge = BottomLeft;
	}
	else if (onBottomRight) {
		_edge = BottomRight;
	}
	else if (onTopRight) {
		_edge = TopRight;
	}
	else if (onTopLeft) {
		_edge = TopLeft;
	}
	else {
		_edge = None;
	}
}

void Widget::setRadius(const qreal &radius) {
	_radius = radius;
}

qreal Widget::radius() const {
	return _radius;
}

void Widget::setBorderWidth(const qint16 &borderWidth) {
	_borderWidth = borderWidth;
}

qint16 Widget::borderWidth() const {
	return _borderWidth;
}

File: Ui/PyQt5.3.2/CustomWindow/Widget.py
from PyQt5.QtWidgets import *
from PyQt5.QtCore import *
from PyQt5.QtGui import *


class Edge(object):
    NoEdge = 0x0
    Left = 0x1
    Right = 0x2
    Top = 0x3
    Bottom = 0x4
    TopLeft = 0x5
    TopRight = 0x6
    BottomLeft = 0x7
    BottomRight = 0x8


class Widget(QMainWindow):
    def __init__(self, borderwidth = None, radius = None, *args, **kwargs):
        super(Widget, self).__init__(*args, **kwargs)
        self._cursorChanged = False
        self._leftButtonPressed = False
        self._rubberBand = QRubberBand(QRubberBand.Rectangle)
        self._mouseMove = Edge.NoEdge
        self._mousePress = Edge.NoEdge
        if borderwidth:
            self._borderWidth = borderwidth
        else:
            self._borderWidth = 5
        if radius:
            self._radius = radius
        else:
            self._radius = 5.0
        self.setWindowFlags(Qt.FramelessWindowHint | Qt.CustomizeWindowHint)
        self.setAttribute(Qt.WA_Hover)
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.setMouseTracking(True)
        self.installEventFilter(self)

    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setBrush(QColor("#FFFFFF"))
        painter.setPen(Qt.NoPen)
        painter.setRenderHint(QPainter.Antialiasing)

        painterpath = QPainterPath()
        painterpath.addRoundedRect(QRectF(0, 0, self.width(), self.height()), self._radius, self._radius)
        painter.drawPath(painterpath.simplified())

    def eventFilter(self, obj, event):
        if event.type() == QEvent.MouseMove:
            self.__mouseMove(QMouseEvent(event))
            return True
        else:
            if event.type() == QEvent.Leave:
                self.__mouseLeave(event)
                return True
            else:
                if event.type() == QEvent.HoverMove:
                    self.__mouseHover(QHoverEvent(event))
                    return True
                else:
                    if event.type() == QEvent.MouseButtonPress:
                        self.__mousePress(QMouseEvent(event))
                        return True
                    else:
                        if event.type() == QEvent.MouseButtonRelease:
                            self.__mouseRelease(QMouseEvent(event))
                            return True
                        else:
                            return False

    def __mouseHover(self, event):
        self.__updateCursorShape(self.mapToGlobal(event.pos()))

    def __mouseLeave(self, event):
        if self._leftButtonPressed:
            self.unsetCursor()

    def __mousePress(self, event):
        if event.button() == Qt.LeftButton:
            self._leftButtonPressed = True
            self._mousePress = self.__calculateCursorPosition(event.globalPos(), self.frameGeometry())
            if self._mousePress != Edge.NoEdge:
                self._rubberBand.setGeometry(self.frameGeometry())

    def __mouseRelease(self, event):
        if event.button() == Qt.LeftButton:
            self._leftButtonPressed = False

    def __mouseMove(self, event):
        if self._leftButtonPressed:
            if self._mousePress != Edge.NoEdge:
                originalRect = self._rubberBand.frameGeometry()
                left = originalRect.left()
                top = originalRect.top()
                right = originalRect.right()
                bottom = originalRect.bottom()
                if self._mousePress == Edge.Top:
                    top = event.globalPos().y()
                else:
                    if self._mousePress == Edge.Bottom:
                        bottom = event.globalPos().y()
                    else:
                        if self._mousePress == Edge.Right:
                            right = event.globalPos().x()
                        else:
                            if self._mousePress == Edge.Left:
                                left = event.globalPos().x()
                            else:
                                if self._mousePress == Edge.TopRight:
                                    top = event.globalPos().y()
                                    right = event.globalPos().x()
                                else:
                                    if self._mousePress == Edge.TopLeft:
                                        top = event.globalPos().y()
                                        left = event.globalPos().x()
                                    else:
                                        if self._mousePress == Edge.BottomLeft:
                                            bottom = event.globalPos().y()
                                            left = event.globalPos().x()
                                        else:
                                            if self._mousePress == Edge.BottomRight:
                                                bottom = event.globalPos().y()
                                                right = event.globalPos().x()

                newrect = QRect(QPoint(left, top), QPoint(right, bottom))
                if newrect.width() < self.minimumWidth():
                    left = self.frameGeometry().x()
                else:
                    if newrect.height() < self.minimumHeight():
                        top = self.frameGeometry().y()
                self.setGeometry(QRect(QPoint(left, top), QPoint(right, bottom)))
                self._rubberBand.setGeometry(QRect(QPoint(left, top), QPoint(right, bottom)))

    def __updateCursorShape(self, pos):
        if self.isFullScreen() or self.isMaximized():
            if self._cursorChanged:
                self.unsetCursor()
                return
        self._mouseMove = self.__calculateCursorPosition(pos, self.frameGeometry())
        if self._mouseMove == Edge.Left or self._mouseMove == Edge.Right:
            self.setCursor(Qt.SizeHorCursor)
            self._cursorChanged = True
        else:
            if self._mouseMove == Edge.Top or self._mouseMove == Edge.Bottom:
                self.setCursor(Qt.SizeVerCursor)
                self._cursorChanged = True
            else:
                if self._mouseMove == Edge.TopLeft or self._mouseMove == Edge.BottomRight:
                    self.setCursor(Qt.SizeFDiagCursor)
                    self._cursorChanged = True
                else:
                    if self._mouseMove == Edge.TopRight or self._mouseMove == Edge.BottomLeft:
                        self.setCursor(Qt.SizeBDiagCursor)
                        self._cursorChanged = True
                    else:
                        self.unsetCursor()
                        self._cursorChanged = False

    def __updateRubberBand(self):
        if not self._rubberBand:
            self._rubberBand = QRubberBand(QRubberBand.Rectangle)
            self._rubberBand.show()
        else:
            self._rubberBand = False

    def __calculateCursorPosition(self, pos, framerect):
        onLeft = pos.x() <= framerect.x() + self._borderWidth and \
                 pos.x() >= framerect.x() and \
                 pos.y() <= framerect.y() + framerect.height() - self._borderWidth and \
                 pos.y() >= framerect.y() + self._borderWidth

        onRight = pos.x() >= framerect.x() + framerect.width() - self._borderWidth and \
                  pos.x() <= framerect.x() + framerect.width() and \
                  pos.y() >= framerect.y() + self._borderWidth and \
                  pos.y() <= framerect.y() + framerect.height() - self._borderWidth - 2

        onBottom = pos.x() >= framerect.x() + self._borderWidth and \
                   pos.x() <= framerect.x() + framerect.width() - self._borderWidth - 2 and \
                   pos.y() >= framerect.y() + framerect.height() - self._borderWidth and \
                   pos.y() <= framerect.y() + framerect.height()

        onTop = pos.x() >= framerect.x() + self._borderWidth and \
                pos.x() <= framerect.x() + framerect.width() - self._borderWidth and \
                pos.y() >= framerect.y() and pos.y() <= framerect.y() + self._borderWidth

        onBottomLeft = pos.x() <= framerect.x() + self._borderWidth and \
                       pos.x() >= framerect.x() and \
                       pos.y() <= framerect.y() + framerect.height() and \
                       pos.y() >= framerect.y() + framerect.height() - self._borderWidth

        onBottomRight = pos.x() >= framerect.x() + framerect.width() - self._borderWidth and \
                        pos.x() <= framerect.x() + framerect.width() and \
                        pos.y() >= framerect.y() + framerect.height() - self._borderWidth and \
                        pos.y() <= framerect.y() + framerect.height()

        onTopRight = pos.x() >= framerect.x() + framerect.width() - self._borderWidth and \
                     pos.x() <= framerect.x() + framerect.width() and \
                     pos.y() >= framerect.y() and \
                     pos.y() <= framerect.y() + self._borderWidth

        onTopLeft = pos.x() >= framerect.x() and \
                    pos.x() <= framerect.x() + self._borderWidth and \
                    pos.y() >= framerect.y() and \
                    pos.y() <= framerect.y() + self._borderWidth

        if onLeft:
            return Edge.Left
        else:
            if onRight:
                return Edge.Right
            else:
                if onBottom:
                    return Edge.Bottom
                else:
                    if onTop:
                        return Edge.Top
                    else:
                        if onBottomLeft:
                            return Edge.BottomLeft
                        else:
                            if onBottomRight:
                                return Edge.BottomRight
                            else:
                                if onBottomRight:
                                    return Edge.BottomRight
                                else:
                                    if onTopRight:
                                        return Edge.TopRight
                                    else:
                                        if onTopLeft:
                                            return Edge.TopLeft
                                        else:
                                            return Edge.NoEdge

    def setBorderWidth(self, borderWidth):
        self._borderWidth = borderWidth

    def borderWidth(self):
        return self._borderWidth

    def setRadius(self, radius):
        self._radius = radius

    def radius(self):
        return self.radius()
